File: src/frontend/src/lib/types/data_source.ts
export interface DataSourceConfig {
  source_id: string;
  name: string;
  enabled: boolean;
  api_key?: string | null;
  priority: number;
  requires_api_key: boolean;
  supports_search: boolean;
  extra_config?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface DataSourceConfigUpdate {
  enabled?: boolean;
  api_key?: string | null;
  priority?: number;
  extra_config?: Record<string, unknown> | null;
}

export interface DataSourcePriorityItem {
  source_id: string;
  priority: number;
}

export interface DataSourceTestResult {
  source_id: string;
  success: boolean;
  message: string;
  latency_ms?: number | null;
}

/** Extended ETF search result that includes the source annotation */
export interface ETFSearchResultWithSource {
  symbol: string;
  shortName?: string | null;
  longName?: string | null;
  currency?: string | null;
  isin?: string | null;
  exchange?: string | null;
  fund_family?: string | null;
  category?: string | null;
  source: string;
  source_name: string;
  symbol_derived?: boolean;
}

File: src/backend/app/schemas/data_source.py
from pydantic import BaseModel
from typing import Optional, Any
from datetime import datetime


class DataSourceConfigBase(BaseModel):
    name: str
    enabled: bool = True
    api_key: Optional[str] = None
    priority: int = 100
    requires_api_key: bool = False
    supports_search: bool = False
    extra_config: Optional[Any] = None


class DataSourceConfigResponse(DataSourceConfigBase):
    source_id: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DataSourceConfigUpdate(BaseModel):
    enabled: Optional[bool] = None
    api_key: Optional[str] = None
    priority: Optional[int] = None
    extra_config: Optional[Any] = None


class DataSourcePriorityItem(BaseModel):
    source_id: str
    priority: int


class DataSourcePrioritiesUpdate(BaseModel):
    priorities: list[DataSourcePriorityItem]


class DataSourceTestResult(BaseModel):
    source_id: str
    success: bool
    message: str
    latency_ms: Optional[int] = None


class ETFSourceSymbolResponse(BaseModel):
    id: int
    etf_id: str
    source_id: str
    symbol: str
    verified: bool
    last_verified_at: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}
